File: 2015/24/rust/src/main.rs
use itertools::Itertools;
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";

// --- Find the smalles Quantum Entanglement ---
fn quantum_entanglement(list: &Vec<usize>, groups: usize) -> usize {
    let target = list.iter().sum::<usize>() / groups;
    let mut min = std::usize::MAX;

    for i in 1 .. list.len() {
        for v in list.iter().combinations(i) {
            if v.iter().map(|x| **x).sum::<usize>() == target {
                min = std::cmp::min(min, v.iter().map(|x| **x).product())
            }
        }
        if min != std::usize::MAX { return min; }
    }
    panic!("Not found!");
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    let list: Vec<_> = BufReader::new(file).lines()
        .map(|line| line.unwrap().parse::<usize>().unwrap())
        .collect();

    // --- Puzzle 1 ---
    let qe = quantum_entanglement(&list, 3);
    println!("1. Quantum entaglement: {}", qe);

    // --- Puzzle 2 ---
    let qe = quantum_entanglement(&list, 4);
    println!("1. Quantum entaglement: {}", qe);

    Ok(())
}

File: 2015/24/python/main.py
from functools import reduce
from itertools import combinations

# --- Constants --
INPUT_FILE = "../input/input.txt"

# --- Functions ---
def quantum_entanglement(ls, groups):
    target = sum(ls) / groups
    _min = float("inf")

    for i in range(1, len(ls)):
        for c in combinations(ls, i):
            if sum(c) == target:
                _min = min(_min, reduce(lambda a, b: a * b, c))
        if _min != float("inf"): return _min

    raise Exception("Not found!")

# --- Variables ---
ls = []

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
while True:
    line = file.readline()
    if not line: break
    line = line.strip()

    ls.append(int(line))
file.close()

# --- Puzzle 1 ---
qe = quantum_entanglement(ls, 3)
print(f"1. Quantum entanglement: {qe:,d}")

# --- Puzzle 2 ---
qe = quantum_entanglement(ls, 4)
print(f"2. Quantum entanglement: {qe:,d}")
